feat: Add new stdlib function to_json (#2860)
grafana/alloy: docs/sources/reference/stdlib/encoding.md

c3RyaW5nMTIzIT8kKiYoKSctPUB-
```

## encoding.to_json

The `encoding.to_json` function encodes the map into a JSON string.

> `encoding.to_json` fails if the input argument provided can't be parsed as JSON string.

Reviewer: ```suggestion
`encoding.to_json` fails if the input argument provided can't be parsed as a JSON string.
```

> A common use case for `encoding.to_json` is to encode a configuration of component which is expected to be a JSON string

Reviewer: ```suggestion
A common use case for `encoding.to_json` is to encode a configuration of component which is expected to be a JSON string.
```

> For example, `config` param of [`prometheus.exporter.blackbox`][]

Reviewer: ```suggestion
For example, `config` param of [`prometheus.exporter.blackbox`][].
```

I'm not sure what "`config` param" means in this context. 

Can this be reworked as:

For example, a configuration parameter in [`prometheus.exporter.blackbox`][].

Does this mean the same thing?

Author: `prometheus.exporter.blackbox` component accepts a argument called `config` which is expected to be a JSON string.  May be changing the word param can be more appropriate I believe. https://grafana.com/docs/alloy/latest/reference/components/prometheus/prometheus.exporter.blackbox/#arguments

c3RyaW5nMTIzIT8kKiYoKSctPUB-
```

## encoding.to_json

> The `encoding.to_json` function encodes a map into a JSON.

Reviewer 2: Thanks for your contribution, the code looks good but it looks like it can handle other types that map. Could you add tests for the different [types](https://grafana.com/docs/alloy/latest/get-started/configuration-syntax/expressions/types_and_values/) and document the ones that don't work? (Not sure what happens if you try to encode a Secret or a Capsule for example)

Author: I have updated the tests and documentation. For Secret and Capsule the function converts the public fields of the underlying struct. I updated the doc with the closest understandable text. Should we handle this differently ?

Example configuration the I tried for verifying the behaviour,
```
discovery.relabel "keep_backend_only" {
  targets = [
    { "__meta_foo" = "foo", "__address__" = "localhost", "instance" = "ones",   "app" = "backend" },
    { "__meta_bar" = "bar", "__address__" = "localhost", "instance" = "two",   "app" = "database" },
    { "__meta_baz" = "baz", "__address__" = "localhost", "instance" = "three", "app" = "frontend" },
  ]

  rule {
    source_labels = ["__address__", "instance"]
    separator     = "/"
    target_label  = "destination"
    action        = "replace"
  }

  rule {
    source_labels = ["app"]
    action        = "keep"
    regex         = "backend"
  }
}

declare "discovery_result" {
  argument "output_value" {
    optional = false
  }

  argument "rules_value" {
    optional = false
  }

  argument "number_value" {
    optional = true
    default = 14
  }

  argument "array_value" {
    optional = true
  }

  argument "string_value" {
    optional = true
  }

  argument "null_value" {
    optional = true
  }
}

discovery_result "backend" {
  output_value = encoding.to_json(discovery.relabel.keep_backend_only.output)
  rules_value = encoding.to_json(discovery.relabel.keep_backend_only.rules)
  number_value = encoding.to_json(23)
  array_value = encoding.to_json([1,2,3])
  string_value = encoding.to_json("Hello World")
  null_value = encoding.to_json(null)
}
```

Reviewer 2: Thanks for testing it, I think that the best approach would be to prevent capsule and secret/optional secrets from being converted because the only way to expose a secret should be through the nonsensitive function + I don't see a good usage to get the public fields exported of the capsule for now.

Author: I made the changes to support only map type instead of complicating the support for all other types.

The way i see it,
1. we can add conditions on top of map[string]interface{} for other types int, float, string etc,
2. We can have a recursive logic to convert the applicable types [types](https://grafana.com/docs/alloy/latest/get-started/configuration-syntax/expressions/types_and_values/) since a map can have have nested maps and arrays can have map nested.

Let me know your thoughts or if we have any other approaches here ?

Reviewer 2: good idea, I don't think that the recursion is needed for now, let's keep it simple and extend it if needed in the future

> The `encoding.to_json` function encodes a map into a JSON.

Reviewer: ```suggestion
The `encoding.to_json` function encodes a map into a JSON string.
```

Is string the right word here?

> `encoding.to_json` fails if the map argument provided can't be parsed as json string.

Reviewer: ```suggestion
`encoding.to_json` fails if the map argument provided can't be parsed as JSON string.
```

> A common use case of `encoding.to_json` is to encode a config of component which is expected to be a JSON

Reviewer: ```suggestion
A common use case for `encoding.to_json` is to encode a configuration of component which is expected to be a JSON string.
```

JSON string?

> For example, `config` of a [`prometheus.exporter.blackbox`][]

Reviewer: ```suggestion
For example, configuration of a [`prometheus.exporter.blackbox`][]
```

Missing a word here... a `prometheus.exporter.blackbox` string?

Author: I wanted to refer the param name in the component updated accordingly.

### Examples

```alloy

> > encoding.to_json({"key"={"key1"="val1"}, "key2"=2})
> {"foo":{"foo1":"Hello!"},"key":12}

Reviewer: ```suggestion
> encoding.to_json({"key"={"key1"="val1"}, "key2"=2})
{"example":{"example1":"Hello!"},"key":12}
```

We should avoid using `foo`, `bar`, and `baz` in the documentation. If `example` isn't appropriate here, then use some other generic placeholder.  (I'm working on removing foo, bar, baz from other places in the docs)

> 

Reviewer 2: could you add an example with prometheus.exporter.blackbox please?
